kv/kvclient/kvstreamer: TestStreamerBudgetErrorInEnqueue failed

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4132347&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4132347&tab=artifacts#/) on master @ [5ad21e3896ee809e9c3ebc28bb22166f1275acca](https://github.com/cockroachdb/cockroach/commits/5ad21e3896ee809e9c3ebc28bb22166f1275acca):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 1048627 [recovered]
	panic: root: no bytes in account to release, current 10, free 1048627

```
Stack: 

```
goroutine 2080 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0015303f0, {0x4e6b130, 0xc0014076a0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3acbfe0, 0xc000d0ad38})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4e6b130, 0xc0014076a0}, 0xc0005d4000, {0x3e35ec2, 0x40}, {0xc0021f3a18, 0x7fbdc8ed1108, 0x50})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc000d0a4f8, {0x4e6b130, 0xc0014076a0}, 0x100033)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc001d1cc80, {0x4e6b130, 0xc0014076a0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xa9
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4e6b130, 0xc0014076a0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x3af
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/single_key_exceeds_root_pool_size
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- GOFLAGS=-parallel=4
</p>
</details>
/cc @cockroachdb/kv
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4132348&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4132348&tab=artifacts#/) on master @ [5ad21e3896ee809e9c3ebc28bb22166f1275acca](https://github.com/cockroachdb/cockroach/commits/5ad21e3896ee809e9c3ebc28bb22166f1275acca):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 66 [recovered]
	panic: root: no bytes in account to release, current 10, free 66

```
Stack: 

```
goroutine 132737 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc001864630, {0x4e75650, 0xc003a4b0a0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3ad3680, 0xc00166d578})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4e75650, 0xc003a4b0a0}, 0xc003894000, {0x3e3cee3, 0x5}, {0xc0040f3a08, 0x387fd00, 0xc001354fc0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc0021cabe8, {0x4e75650, 0xc003a4b0a0}, 0x42)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc001354fc0, {0x4e75650, 0xc003a4b0a0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xb0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4e75650, 0xc003a4b0a0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x3af
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/multiple_keys_exceed_limit
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4135360&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4135360&tab=artifacts#/) on master @ [e84001d8611aedc45a413019f5e6317c2b757ca0](https://github.com/cockroachdb/cockroach/commits/e84001d8611aedc45a413019f5e6317c2b757ca0):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 66 [recovered]
	panic: root: no bytes in account to release, current 10, free 66

```
Stack: 

```
goroutine 133006 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000d6aab0, {0x4e746b0, 0xc0032d3ca0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3ad2c00, 0xc0049130b0})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4e746b0, 0xc0032d3ca0}, 0xc003dd0000, {0x3e3beeb, 0x5}, {0xc004165a08, 0x387ef60, 0xc000886f80})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc0043c6828, {0x4e746b0, 0xc0032d3ca0}, 0x42)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc000886f80, {0x4e746b0, 0xc0032d3ca0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xb0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4e746b0, 0xc0032d3ca0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x3af
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/multiple_keys_exceed_limit
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4141052&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4141052&tab=artifacts#/) on master @ [365b4da8bd02c06ee59d2130a56dec74ffc9ce21](https://github.com/cockroachdb/cockroach/commits/365b4da8bd02c06ee59d2130a56dec74ffc9ce21):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 66 [recovered]
	panic: root: no bytes in account to release, current 10, free 66

```
Stack: 

```
goroutine 164655 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000423320, {0x4e7b830, 0xc00434b9a0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3ad84e0, 0xc000ae8048})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4e7b830, 0xc00434b9a0}, 0xc005176000, {0x3e41e35, 0x5}, {0xc0042ffa08, 0x3884480, 0xc002270e40})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc002a69a40, {0x4e7b830, 0xc00434b9a0}, 0x42)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc002270e40, {0x4e7b830, 0xc00434b9a0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xb0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4e7b830, 0xc00434b9a0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x3af
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:488 +0x12a
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:478 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/multiple_keys_exceed_limit
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4173083&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4173083&tab=artifacts#/) on master @ [58ceac139a7e83052171121b28026a7366f16f7e](https://github.com/cockroachdb/cockroach/commits/58ceac139a7e83052171121b28026a7366f16f7e):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 66 [recovered]
	panic: root: no bytes in account to release, current 10, free 66

```
Stack: 

```
goroutine 168712 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000d75050, {0x4eb0df8, 0xc004370da0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3afa940, 0xc0034b68a0})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4eb0df8, 0xc004370da0}, 0xc0056ba000, {0x3e66d8b, 0x5}, {0xc0057539d8, 0x38a4be0, 0xc003036ee0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc004388af8, {0x4eb0df8, 0xc004370da0}, 0x42)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc003036ee0, {0x4eb0df8, 0xc004370da0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xb0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4eb0df8, 0xc004370da0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x36f
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x16f
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/multiple_keys_exceed_limit
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4190387&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4190387&tab=artifacts#/) on master @ [8cd28089c6c7333615ba3201e841839001d2f0e1](https://github.com/cockroachdb/cockroach/commits/8cd28089c6c7333615ba3201e841839001d2f0e1):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 1048627 [recovered]
	panic: root: no bytes in account to release, current 10, free 1048627

```
Stack: 

```
goroutine 187580 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc000c5ec60, {0x4ebeab8, 0xc00525fda0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3b13960, 0xc005a76918})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4ebeab8, 0xc00525fda0}, 0xc004f0c000, {0x3e820ef, 0x5}, {0xc005eb19d8, 0x38bce20, 0xc0057fa2e0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc004a6f788, {0x4ebeab8, 0xc00525fda0}, 0x100033)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc0057fa2e0, {0x4ebeab8, 0xc00525fda0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xb0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4ebeab8, 0xc00525fda0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x36f
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x16f
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/single_key_exceeds_root_pool_size
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

kv/kvclient/kvstreamer.TestStreamerBudgetErrorInEnqueue [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=4211895&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=4211895&tab=artifacts#/) on master @ [ae101ea32d99a9142a319b0a1f6850ee76d55cd9](https://github.com/cockroachdb/cockroach/commits/ae101ea32d99a9142a319b0a1f6850ee76d55cd9):

Fatal error:

```
panic: root: no bytes in account to release, current 10, free 1048627 [recovered]
	panic: root: no bytes in account to release, current 10, free 1048627

```
Stack: 

```
goroutine 149361 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Recover(0xc0014345a0, {0x4edbed8, 0xc004c1a3c0})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:251 +0x94
panic({0x3b1fec0, 0xc002bf5ad0})
	/usr/local/go/src/runtime/panic.go:1038 +0x215
github.com/cockroachdb/cockroach/pkg/util/log/logcrash.ReportOrPanic({0x4edbed8, 0xc004c1a3c0}, 0xc0049da000, {0x3e8fb29, 0x5}, {0xc0053b39d8, 0x38c8e60, 0xc003987c60})
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/logcrash/crash_reporting.go:376 +0x1af
github.com/cockroachdb/cockroach/pkg/util/mon.(*BoundAccount).Shrink(0xc000f31ad0, {0x4edbed8, 0xc004c1a3c0}, 0x100033)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/mon/bytes_usage.go:636 +0x127
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*budget).release(0xc003987c60, {0x4edbed8, 0xc004c1a3c0}, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/budget.go:117 +0xb0
github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer.(*workerCoordinator).performRequestAsync.func1({0x4edbed8, 0xc004c1a3c0})
	/go/src/github.com/cockroachdb/cockroach/pkg/kv/kvclient/kvstreamer/streamer.go:1025 +0x36f
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2()
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:494 +0x16f
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx
	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:483 +0x445
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestStreamerBudgetErrorInEnqueue/single_key_exceeds_root_pool_size
```
</p>
</details>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
Parameters in this failure:

- TAGS=deadlock

- GOFLAGS=-parallel=4
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestStreamerBudgetErrorInEnqueue.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/kv/kvclient/kvstreamer/BUILD.bazel
+++ b/pkg/kv/kvclient/kvstreamer/BUILD.bazel
@@ -18,6 +18,7 @@ go_library(
         "//pkg/settings",
         "//pkg/settings/cluster",
         "//pkg/util/admission",
+        "//pkg/util/buildutil",
         "//pkg/util/mon",
         "//pkg/util/quotapool",
         "//pkg/util/stop",

--- a/pkg/kv/kvclient/kvstreamer/BUILD.bazel
+++ b/pkg/kv/kvclient/kvstreamer/BUILD.bazel
@@ -52,6 +52,7 @@ go_test(
         "//pkg/util/log",
         "//pkg/util/mon",
         "//pkg/util/randutil",
+        "@com_github_dustin_go_humanize//:go-humanize",
         "@com_github_stretchr_testify//require",
     ],
 )

--- a/pkg/kv/kvclient/kvstreamer/budget.go
+++ b/pkg/kv/kvclient/kvstreamer/budget.go
@@ -12,6 +12,7 @@ package kvstreamer
 
 import (
 	"context"
+	"sync"
 
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
@@ -23,8 +24,9 @@ import (
 // This struct is a wrapper on top of mon.BoundAccount because we want to
 // support the notion of budget "going in debt". This can occur in a degenerate
 // case when a single large row exceeds the provided limit. The Streamer is
-// expected to have only a single request in flight in this case. Additionally,
-// the budget provides blocking (via waitCh) until it gets out of debt.
+// expected to have only a single request in progress in this case.
+// Additionally, the budget provides blocking until some memory is returned to
+// the budget or, if the budget is already in debt, until it gets out of debt.
 type budget struct {
 	mu struct {
 		// If the Streamer's mutex also needs to be locked, the budget's mutex
@@ -33,15 +35,18 @@ type budget struct {
 		// acc represents the current reservation of this budget against the
 		// root memory pool.
 		acc *mon.BoundAccount
+		// waitForBudget is used by the main loop of the workerCoordinator to
+		// block until the next release() call or, if the budget is currently in
+		// debt, until it gets out of debt.
+		waitForBudget *sync.Cond
 	}
 	// limitBytes is the maximum amount of bytes that this budget should reserve
 	// against acc, i.e. acc.Used() should not exceed limitBytes. However, in a
 	// degenerate case of a single large row, the budget can go into debt and
 	// acc.Used() might exceed limitBytes.
+	//
+	// Available budget can be calculated as limitBytes - mu.acc.Used().
 	limitBytes int64
-	// waitCh is used by the main loop of the workerCoordinator to block until
-	// available() becomes positive (until some release calls occur).
-	waitCh chan struct{}
 }
 
 // newBudget creates a new budget with the specified limit. The limit determines
@@ -58,25 +63,10 @@ type budget struct {
 // to interact with the account only after canceling the Streamer (because
 // memory accounts are not thread-safe).
 func newBudget(acc *mon.BoundAccount, limitBytes int64) *budget {
-	b := budget{
-		limitBytes: limitBytes,
-		waitCh:     make(chan struct{}),
-	}
+	b := budget{limitBytes: limitBytes}
 	b.mu.acc = acc
+	b.mu.waitForBudget = sync.NewCond(&b.mu.Mutex)
 	return &b
-}
-
-// available returns how many bytes are currently available in the budget. The
-// answer can be negative, in case the Streamer has used un-budgeted memory
-// (e.g. one result was very large putting the budget in debt).
-//
-// Note that it's possible that actually available budget is less than the
-// number returned - this might occur if --max-sql-memory root pool is almost
-// used up.
-func (b *budget) available() int64 {
-	b.mu.Lock()
-	defer b.mu.Unlock()
-	return b.limitBytes - b.mu.acc.Used()
 }
 
 // consume draws bytes from the available budget. An error is returned if the
@@ -110,17 +100,22 @@ func (b *budget) consumeLocked(ctx context.Context, bytes int64, allowDebt bool)
 	return b.mu.acc.Grow(ctx, bytes)
 }
 
-// release returns bytes to the available budget.
+// release returns bytes to the available budget. The budget's mutex must not be
+// held.
 func (b *budget) release(ctx context.Context, bytes int64) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
+	b.releaseLocked(ctx, bytes)
+}
+
+// releaseLocked is the same as release but assumes that the budget's mutex is
+// already being held.
+func (b *budget) releaseLocked(ctx context.Context, bytes int64) {
+	b.mu.AssertHeld()
 	b.mu.acc.Shrink(ctx, bytes)
 	if b.limitBytes > b.mu.acc.Used() {
-		// Since we now have some available budget, we non-blockingly send on
-		// the wait channel to notify the mainCoordinator about it.
-		select {
-		case b.waitCh <- struct{}{}:
-		default:
-		}
+		// Since we now have some available budget, signal the worker
+		// coordinator.
+		b.mu.waitForBudget.Signal()
 	}
 }

--- a/pkg/kv/kvclient/kvstreamer/budget.go
+++ b/pkg/kv/kvclient/kvstreamer/budget.go
@@ -16,6 +16,7 @@ import (
 
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/syncutil"
+	"github.com/cockroachdb/errors"
 )
 
 // budget abstracts the memory budget that is provided to the Streamer by its
@@ -94,7 +95,10 @@ func (b *budget) consumeLocked(ctx context.Context, bytes int64, allowDebt bool)
 	// check whether we'll stay within the budget.
 	if !allowDebt && b.limitBytes > 5 {
 		if b.mu.acc.Used()+bytes > b.limitBytes {
-			return mon.MemoryResource.NewBudgetExceededError(bytes, b.mu.acc.Used(), b.limitBytes)
+			return errors.Wrap(
+				mon.MemoryResource.NewBudgetExceededError(bytes, b.mu.acc.Used(), b.limitBytes),
+				"streamer budget",
+			)
 		}
 	}
 	return b.mu.acc.Grow(ctx, bytes)

--- a/pkg/kv/kvclient/kvstreamer/streamer.go
+++ b/pkg/kv/kvclient/kvstreamer/streamer.go
@@ -229,6 +229,10 @@ type Streamer struct {
 		// TODO(yuzefovich): consider using ring.Buffer instead of a slice.
 		requestsToServe []singleRangeBatch
 
+		// hasWork is used by the worker coordinator to block until some
+		// requests are added to requestsToServe.
+		hasWork *sync.Cond
+
 		// numRangesLeftPerScanRequest tracks how many ranges a particular
 		// originally enqueued ScanRequest touches, but scanning of those ranges
 		// isn't complete. It is allocated lazily when the first ScanRequest is
@@ -254,7 +258,14 @@ type Streamer struct {
 		// results are the results of already completed requests that haven't
 		// been returned by GetResults() yet.
 		results []Result
-		err     error
+
+		// err is set once the Streamer cannot make progress any more and will
+		// be returned by GetResults().
+		err error
+
+		// done is set to true once the Streamer is closed meaning the worker
+		// coordinator must exit.
+		done bool
 	}
 }
 
@@ -303,6 +314,7 @@ func NewStreamer(
 		stopper:    stopper,
 		budget:     newBudget(acc, limitBytes),
 	}
+	s.mu.hasWork = sync.NewCond(&s.mu.Mutex)
 	s.coordinator = workerCoordinator{
 		s:                      s,
 		txn:                    txn,
@@ -316,7 +328,6 @@ func NewStreamer(
 		"single Streamer async concurrency",
 		uint64(streamerConcurrencyLimit.Get(&st.SV)),
 	)
-	s.coordinator.mu.hasWork = sync.NewCond(&s.coordinator.mu)
 	streamerConcurrencyLimit.SetOnChange(&st.SV, func(ctx context.Context) {
 		s.coordinator.asyncSem.UpdateCapacity(uint64(streamerConcurrencyLimit.Get(&st.SV)))
 	})
@@ -522,7 +533,7 @@ func (s *Streamer) Enqueue(
 
 	// Memory reservation was approved, so the requests are good to go.
 	s.mu.requestsToServe = requestsToServe
-	s.coordinator.mu.hasWork.Signal()
+	s.mu.hasWork.Signal()
 	return nil
 }
 
@@ -604,6 +615,13 @@ func (s *Streamer) notifyGetResultsLocked() {
 func (s *Streamer) setError(err error) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
+	s.setErrorLocked(err)
+}
+
+// setErrorLocked is the same as setError but assumes that the mutex of s is
+// already being held.
+func (s *Streamer) setErrorLocked(err error) {
+	s.mu.AssertHeld()
 	if s.mu.err == nil {
 		s.mu.err = err
 	}
@@ -616,11 +634,11 @@ func (s *Streamer) setError(err error) {
 func (s *Streamer) Close() {
 	if s.coordinatorStarted {
 		s.coordinatorCtxCancel()
-		s.coordinator.mu.Lock()
-		s.coordinator.mu.done = true
+		s.mu.Lock()
+		s.mu.done = true
 		// Unblock the coordinator in case it is waiting for more work.
-		s.coordinator.mu.hasWork.Signal()
-		s.coordinator.mu.Unlock()
+		s.mu.hasWork.Signal()
+		s.mu.Unlock()
 	}
 	s.waitGroup.Wait()
 	*s = Streamer{}
@@ -687,14 +705,6 @@ type workerCoordinator struct {
 	// For request and response admission control.
 	requestAdmissionHeader roachpb.AdmissionHeader
 	responseAdmissionQ     *admission.WorkQueue
-
-	mu struct {
-		syncutil.Mutex
-		hasWork *sync.Cond
-		// done is set to true once the Streamer is closed meaning the worker
-		// coordinator must exit.
-		done bool
-	}
 }
 
 // mainLoop runs throughout the lifetime of the Streamer (from the first Enqueue
@@ -707,23 +717,17 @@ func (w *workerCoordinator) mainLoop(ctx context.Context) {
 	defer w.s.waitGroup.Done()
 	for {
 		// Get next requests to serve.
-		requestsToServe, avgResponseSize, shouldExit := w.getRequests()
+		requestsToServe, avgResponseSize, shouldExit := w.getRequests(ctx)
 		if shouldExit {
 			return
 		}
-		if len(requestsToServe) == 0 {
-			// If the Streamer isn't closed yet, block until there are enqueued
-			// requests.
-			w.mu.Lock()
-			if !w.mu.done {
-				w.mu.hasWork.Wait()
+		if buildutil.CrdbTestBuild {
+			if len(requestsToServe) == 0 {
+				panic(errors.AssertionFailedf(
+					"unexpectedly zero requests to serve returned by " +
+						"getRequests when shouldExit is false",
+				))
 			}
-			w.mu.Unlock()
-			if ctx.Err() != nil {
-				w.s.setError(ctx.Err())
-				return
-			}
-			continue
 		}
 
 		// Now wait until there is enough budget to at least receive one full
@@ -748,19 +752,52 @@ func (w *workerCoordinator) mainLoop(ctx context.Context) {
 	}
 }
 
-// getRequests returns all currently enqueued requests to be served.
+// getRequests returns all currently enqueued requests to be served, and if
+// there are none, it waits until some requests are added. It assumes that the
+// Streamer's mutex is not held.
 //
 // A boolean that indicates whether the coordinator should exit is returned.
-func (w *workerCoordinator) getRequests() (
+func (w *workerCoordinator) getRequests(
+	ctx context.Context,
+) (requestsToServe []singleRangeBatch, avgResponseSize int64, shouldExit bool) {
+	w.s.mu.Lock()
+	defer w.s.mu.Unlock()
+	requestsToServe, avgResponseSize, shouldExit = w.getRequestsLocked()
+	if shouldExit {
+		return requestsToServe, avgResponseSize, shouldExit
+	}
+	if len(requestsToServe) == 0 {
+		w.s.mu.hasWork.Wait()
+		// Check whether the Streamer has been canceled or closed while we
+		// were waiting for work.
+		if ctx.Err() != nil {
+			// Note that if it was a graceful closure, we still will see a
+			// context cancellation error; however, it won't be propagated
+			// to the client because no more calls after Close() are
+			// allowed.
+			w.s.setErrorLocked(ctx.Err())
+			shouldExit = true
+			return
+		}
+		requestsToServe, avgResponseSize, shouldExit = w.getRequestsLocked()
+	}
+	return requestsToServe, avgResponseSize, shouldExit
+}
+
+// getRequestsLocked returns all currently enqueued requests to be served.
+// Unlike getRequests, it doesn't block if there are no requests at the moment.
+// It assumes that the Streamer's mutex is being held.
+//
+// A boolean that indicates whether the coordinator should exit is returned.
+func (w *workerCoordinator) getRequestsLocked() (
 	requestsToServe []singleRangeBatch,
 	avgResponseSize int64,
 	shouldExit bool,
 ) {
-	w.s.mu.Lock()
-	defer w.s.mu.Unlock()
+	w.s.mu.AssertHeld()
 	requestsToServe = w.s.mu.requestsToServe
 	avgResponseSize = w.s.mu.avgResponseEstimator.getAvgResponseSize()
-	shouldExit = w.s.mu.err != nil
+	shouldExit = w.s.mu.err != nil || w.s.mu.done
 	return requestsToServe, avgResponseSize, shouldExit
 }
 
@@ -873,7 +910,7 @@ func (w *workerCoordinator) addRequest(req singleRangeBatch) {
 	w.s.mu.Lock()
 	defer w.s.mu.Unlock()
 	w.s.mu.requestsToServe = append(w.s.mu.requestsToServe, req)
-	w.mu.hasWork.Signal()
+	w.s.mu.hasWork.Signal()
 }
 
 func (w *workerCoordinator) asyncRequestCleanup() {

--- a/pkg/kv/kvclient/kvstreamer/streamer.go
+++ b/pkg/kv/kvclient/kvstreamer/streamer.go
@@ -97,9 +97,9 @@ type Result struct {
 	// requests.
 	EnqueueKeysSatisfied []int
 	// memoryTok describes the memory reservation of this Result that needs to
-	// be released back to the budget when the Result is Release()'d.
+	// be released back to the Streamer's budget when the Result is Release()'d.
 	memoryTok struct {
-		budget    *budget
+		streamer  *Streamer
 		toRelease int64
 	}
 	// position tracks the ordinal among all originally enqueued requests that
@@ -131,8 +131,14 @@ type Hints struct {
 // Streamer internally does buffering and caching of Results - which also
 // contributes to the refcounts.
 func (r Result) Release(ctx context.Context) {
-	if r.memoryTok.budget != nil {
-		r.memoryTok.budget.release(ctx, r.memoryTok.toRelease)
+	if s := r.memoryTok.streamer; s != nil {
+		s.budget.mu.Lock()
+		defer s.budget.mu.Unlock()
+		s.budget.releaseLocked(ctx, r.memoryTok.toRelease)
+		s.mu.Lock()
+		defer s.mu.Unlock()
+		s.mu.numUnreleasedResults--
+		s.signalBudgetIfNoRequestsInProgressLocked()
 	}
 }
 
@@ -251,9 +257,13 @@ type Streamer struct {
 		// are currently being served asynchronously (i.e. those that have
 		// already left requestsToServe queue, but for which we haven't received
 		// the results yet).
-		// TODO(yuzefovich): check whether the contention on mu when accessing
-		// this field is sufficient to justify pulling it out into an atomic.
 		numRequestsInFlight int
+
+		// numUnreleasedResults tracks the number of Results that have already
+		// been created but haven't been Release()'d yet. This number includes
+		// those that are currently in 'results' in addition to those returned
+		// by GetResults() to the caller which the caller hasn't processed yet.
+		numUnreleasedResults int
 
 		// results are the results of already completed requests that haven't
 		// been returned by GetResults() yet.
@@ -639,17 +649,36 @@ func (s *Streamer) Close() {
 		// Unblock the coordinator in case it is waiting for more work.
 		s.mu.hasWork.Signal()
 		s.mu.Unlock()
+		// Unblock the coordinator in case it is waiting for the budget.
+		s.budget.mu.waitForBudget.Signal()
 	}
 	s.waitGroup.Wait()
 	*s = Streamer{}
 }
 
-// getNumRequestsInFlight returns the number of requests that are currently in
-// flight. This method should be called without holding the lock of s.
-func (s *Streamer) getNumRequestsInFlight() int {
+// getNumRequestsInProgress returns the number of requests that are currently
+// "in progress" - already issued requests that are in flight combined with the
+// number of unreleased results. This method should be called without holding
+// the lock of s.
+func (s *Streamer) getNumRequestsInProgress() int {
 	s.mu.Lock()
 	defer s.mu.Unlock()
-	return s.mu.numRequestsInFlight
+	return s.mu.numRequestsInFlight + s.mu.numUnreleasedResults
+}
+
+// signalBudgetIfNoRequestsInProgressLocked checks whether there are no requests
+// in progress and signals the budget's condition variable if so.
+//
+// We have to explicitly signal the condition variable to make sure that if the
+// budget doesn't get out of debt, the worker coordinator doesn't wait for any
+// other request to be completed / other result be released.
+//
+// The mutex of s must be held.
+func (s *Streamer) signalBudgetIfNoRequestsInProgressLocked() {
+	s.mu.AssertHeld()
+	if s.mu.numRequestsInFlight == 0 && s.mu.numUnreleasedResults == 0 {
+		s.budget.mu.waitForBudget.Signal()
+	}
 }
 
 // adjustNumRequestsInFlight updates the number of requests that are currently
@@ -658,6 +687,7 @@ func (s *Streamer) adjustNumRequestsInFlight(delta int) {
 	s.mu.Lock()
 	defer s.mu.Unlock()
 	s.mu.numRequestsInFlight += delta
+	s.signalBudgetIfNoRequestsInProgressLocked()
 }
 
 // singleRangeBatch contains parts of the originally enqueued requests that have
@@ -730,18 +760,9 @@ func (w *workerCoordinator) mainLoop(ctx context.Context) {
 			}
 		}
 
-		// Now wait until there is enough budget to at least receive one full
-		// response (but only if there are requests in flight - if there are
-		// none, then we might have a degenerate case when a single row is
-		// expected to exceed the budget).
-		// TODO(yuzefovich): consider using a multiple of avgResponseSize here.
-		for w.s.getNumRequestsInFlight() > 0 && w.s.budget.available() < avgResponseSize {
-			select {
-			case <-w.s.budget.waitCh:
-			case <-ctx.Done():
-				w.s.setError(ctx.Err())
-				return
-			}
+		shouldExit = w.waitUntilEnoughBudget(ctx, avgResponseSize)
+		if shouldExit {
+			return
 		}
 
 		err := w.issueRequestsForAsyncProcessing(ctx, requestsToServe, avgResponseSize)
@@ -801,11 +822,42 @@ func (w *workerCoordinator) getRequestsLocked() (
 	return requestsToServe, avgResponseSize, shouldExit
 }
 
+// waitUntilEnoughBudget waits until there is enough budget to at least receive
+// one full response.
+//
+// A boolean that indicates whether the coordinator should exit is returned.
+func (w *workerCoordinator) waitUntilEnoughBudget(
+	ctx context.Context, avgResponseSize int64,
+) (shouldExit bool) {
+	w.s.budget.mu.Lock()
+	defer w.s.budget.mu.Unlock()
+	// TODO(yuzefovich): consider using a multiple of avgResponseSize here.
+	for w.s.budget.limitBytes-w.s.budget.mu.acc.Used() < avgResponseSize {
+		// There isn't enough budget at the moment. Check whether there are any
+		// requests in progress.
+		if w.s.getNumRequestsInProgress() == 0 {
+			// We have a degenerate case when a single row is expected to exceed
+			// the budget.
+			return false
+		}
+		// We have to wait for some budget.release() calls.
+		w.s.budget.mu.waitForBudget.Wait()
+		// Check if the Streamer has been canceled or closed while we were
+		// waiting.
+		if ctx.Err() != nil {
+			w.s.setError(ctx.Err())
+			return true
+		}
+	}
+	return false
+}
+
 // issueRequestsForAsyncProcessing iterates over the given requests and issues
 // them to be served asynchronously while there is enough budget available to
 // receive the responses. Once the budget is exhausted, no new requests are
-// issued, the only exception is made for the case when there are no other
-// requests in flight, and in that scenario, a single request will be issued.
+// issued, the only exception is made for the case when there are no requests in
+// progress (both requests in flight as well as unreleased results), and in that
+// scenario, a single request will be issued.
 //
 // It is assumed that requestsToServe is a prefix of w.s.mu.requestsToServe
 // (i.e. it is possible that some other requests have been appended to
@@ -825,16 +877,16 @@ func (w *workerCoordinator) issueRequestsForAsyncProcessing(
 	w.s.budget.mu.Lock()
 	defer w.s.budget.mu.Unlock()
 
-	headOfLine := w.s.getNumRequestsInFlight() == 0
+	headOfLine := w.s.getNumRequestsInProgress() == 0
 	var budgetIsExhausted bool
 	for numRequestsIssued < len(requestsToServe) && !budgetIsExhausted {
 		availableBudget := w.s.budget.limitBytes - w.s.budget.mu.acc.Used()
 		if availableBudget < avgResponseSize {
 			if !headOfLine {
 				// We don't have enough budget available to serve this request,
-				// and there are other requests in flight, so we'll wait for
+				// and there are other requests in progress, so we'll wait for
 				// some of them to finish.
-				break
+				return nil
 			}
 			budgetIsExhausted = true
 			if availableBudget < 1 {
@@ -871,7 +923,8 @@ func (w *workerCoordinator) issueRequestsForAsyncProcessing(
 			// we're holding the budget's mutex. Thus, the error indicates that
 			// the root memory pool has been exhausted.
 			if !headOfLine {
-				// There are some requests in flight, so we'll let them finish.
+				// There are some requests in progress, so we'll let them
+				// finish / be released.
 				//
 				// This is opportunistic behavior where we're hoping that once
 				// other requests are fully processed (i.e. the corresponding
@@ -884,9 +937,9 @@ func (w *workerCoordinator) issueRequestsForAsyncProcessing(
 				// available RAM). Furthermore, if other queries are consuming
 				// all of the root memory pool limit, then the head-of-the-line
 				// request will notice it and will exit accordingly.
-				break
+				return nil
 			}
-			// We don't have any requests in flight, so we'll exit to be safe
+			// We don't have any requests in progress, so we'll exit to be safe
 			// (in order not to OOM the node). Most likely this occurs when
 			// there are concurrent memory-intensive queries which this Streamer
 			// has no control over.
@@ -932,11 +985,11 @@ func (w *workerCoordinator) asyncRequestCleanup() {
 // the memory reservation (according to the Result's footprint) that will be
 // returned back to the budget once Result.Release() is called.
 //
-// headOfLine indicates whether this request is the current head of the line.
-// Head-of-the-line requests are treated specially in a sense that they are
-// allowed to put the budget into debt. The caller is responsible for ensuring
-// that there is at most one asynchronous request with headOfLine=true at all
-// times.
+// headOfLine indicates whether this request is the current head of the line and
+// there are no unreleased Results. Head-of-the-line requests are treated
+// specially in a sense that they are allowed to put the budget into debt. The
+// caller is responsible for ensuring that there is at most one asynchronous
+// request with headOfLine=true at all times.
 func (w *workerCoordinator) performRequestAsync(
 	ctx context.Context, req singleRangeBatch, targetBytes int64, headOfLine bool,
 ) {
@@ -1028,7 +1081,7 @@ func (w *workerCoordinator) performRequestAsync(
 							EnqueueKeysSatisfied: []int{enqueueKey},
 							position:             req.positions[i],
 						}
-						result.memoryTok.budget = w.s.budget
+						result.memoryTok.streamer = w.s
 						result.memoryTok.toRelease = toRelease
 						memoryFootprintBytes += toRelease
 						results = append(results, result)
@@ -1046,7 +1099,7 @@ func (w *workerCoordinator) performRequestAsync(
 							EnqueueKeysSatisfied: []int{enqueueKey},
 							position:             req.positions[i],
 						}
-						result.memoryTok.budget = w.s.budget
+						result.memoryTok.streamer = w.s
 						result.memoryTok.toRelease = toRelease
 						result.ScanResp.ScanResponse = scan
 						// Complete field will be set below.
@@ -1205,6 +1258,7 @@ func (w *workerCoordinator) finalizeSingleRangeResults(
 	// partial one. Think more about this.
 	w.s.mu.avgResponseEstimator.update(actualMemoryReservation, int64(len(results)))
 	w.s.mu.numCompleteRequests += numCompleteResponses
+	w.s.mu.numUnreleasedResults += len(results)
 	// Store the results and non-blockingly notify the Streamer about them.
 	w.s.mu.results = append(w.s.mu.results, results...)
 	w.s.notifyGetResultsLocked()

--- a/pkg/kv/kvclient/kvstreamer/streamer.go
+++ b/pkg/kv/kvclient/kvstreamer/streamer.go
@@ -25,6 +25,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/settings"
 	"github.com/cockroachdb/cockroach/pkg/settings/cluster"
 	"github.com/cockroachdb/cockroach/pkg/util/admission"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/quotapool"
 	"github.com/cockroachdb/cockroach/pkg/util/stop"
@@ -390,22 +391,21 @@ func (s *Streamer) Enqueue(
 	// TODO(yuzefovich): we might want to have more fine-grained lock
 	// acquisitions once pipelining is implemented.
 	s.mu.Lock()
-	locked := true
 	defer func() {
-		if retErr != nil {
-			if !locked {
-				s.mu.Lock()
-				locked = true
-			}
-			if s.mu.err == nil {
-				// Set the error so that mainLoop of the worker coordinator
-				// exits as soon as possible, without issuing any more requests.
-				s.mu.err = retErr
+		// We assume that the Streamer's mutex is held when Enqueue returns.
+		s.mu.AssertHeld()
+		defer s.mu.Unlock()
+		if buildutil.CrdbTestBuild {
+			if s.mu.err != nil {
+				panic(errors.NewAssertionErrorWithWrappedErrf(s.mu.err, "s.mu.err is non-nil"))
 			}
 		}
-		if locked {
-			s.mu.Unlock()
-		}
+		// Set the error (if present) so that mainLoop of the worker coordinator
+		// exits as soon as possible, without issuing any requests. Note that
+		// s.mu.err couldn't have already been set by the worker coordinator or
+		// the asynchronous requests because the worker coordinator starts
+		// issuing the requests only after Enqueue() returns.
+		s.mu.err = retErr
 	}()
 
 	if enqueueKeys != nil && len(enqueueKeys) != len(reqs) {
@@ -511,18 +511,8 @@ func (s *Streamer) Enqueue(
 		}
 	}
 
-	// Release the Streamer's mutex so that there is no overlap with the
-	// budget's mutex - the budget's mutex needs to be acquired first in order
-	// to eliminate a potential deadlock.
-	s.mu.Unlock()
-	locked = false
-
-	// Account for the memory used by all the requests. We allow the budget to
-	// go into debt iff a single request was enqueued. This is needed to support
-	// the case of arbitrarily large keys - the caller is expected to produce
-	// requests with such cases one at a time.
-	allowDebt := len(reqs) == 1
-	if err = s.budget.consume(ctx, totalReqsMemUsage, allowDebt); err != nil {
+	err = s.enqueueMemoryAccountingLocked(ctx, totalReqsMemUsage, len(reqs))
+	if err != nil {
 		return err
 	}
 
@@ -530,6 +520,28 @@ func (s *Streamer) Enqueue(
 	// one singleRangeBatch object has been appended to s.mu.requestsToServe.
 	s.coordinator.mu.hasWork.Signal()
 	return nil
+}
+
+// enqueueMemoryAccountingLocked accounts for the memory used by all the
+// requests that are added in Enqueue.
+//
+// It assumes that the mutex of s is being held and guarantees that it will be
+// held when the function returns too.
+func (s *Streamer) enqueueMemoryAccountingLocked(
+	ctx context.Context, totalReqsMemUsage int64, numReqs int,
+) error {
+	s.mu.AssertHeld()
+	// Per the contract of the budget's mutex (which must be acquired first,
+	// before the Streamer's mutex), we cannot hold the mutex of s, so we have
+	// to release it first and acquire right before returning from this
+	// function.
+	s.mu.Unlock()
+	defer s.mu.Lock()
+	// We allow the budget to go into debt iff a single request was enqueued.
+	// This is needed to support the case of arbitrarily large keys - the caller
+	// is expected to produce requests with such cases one at a time.
+	allowDebt := numReqs == 1
+	return s.budget.consume(ctx, totalReqsMemUsage, allowDebt)
 }
 
 // GetResults blocks until at least one result is available. If the operation

--- a/pkg/kv/kvclient/kvstreamer/streamer.go
+++ b/pkg/kv/kvclient/kvstreamer/streamer.go
@@ -434,6 +434,10 @@ func (s *Streamer) Enqueue(
 	// requestsToServe, and the worker coordinator will then pick those batches
 	// up to execute asynchronously.
 	var totalReqsMemUsage int64
+	// Use a local variable for requestsToServe rather than
+	// s.mu.requestsToServe. This is needed in order for the worker coordinator
+	// to not pick up any work until we account for totalReqsMemUsage.
+	var requestsToServe []singleRangeBatch
 	// TODO(yuzefovich): in InOrder mode we need to treat the head-of-the-line
 	// request differently.
 	seekKey := rs.Key
@@ -492,7 +496,7 @@ func (s *Streamer) Enqueue(
 			sort.Sort(&r)
 		}
 
-		s.mu.requestsToServe = append(s.mu.requestsToServe, r)
+		requestsToServe = append(requestsToServe, r)
 
 		// Determine next seek key, taking potentially sparse requests into
 		// consideration.
@@ -516,8 +520,8 @@ func (s *Streamer) Enqueue(
 		return err
 	}
 
-	// TODO(yuzefovich): it might be better to notify the coordinator once
-	// one singleRangeBatch object has been appended to s.mu.requestsToServe.
+	// Memory reservation was approved, so the requests are good to go.
+	s.mu.requestsToServe = requestsToServe
 	s.coordinator.mu.hasWork.Signal()
 	return nil
 }

--- a/pkg/kv/kvclient/kvstreamer/streamer.go
+++ b/pkg/kv/kvclient/kvstreamer/streamer.go
@@ -738,7 +738,7 @@ type workerCoordinator struct {
 }
 
 // mainLoop runs throughout the lifetime of the Streamer (from the first Enqueue
-// call until Cancel) and routes the single-range batches for asynchronous
+// call until Close) and routes the single-range batches for asynchronous
 // execution. This function is dividing up the Streamer's budget for each of
 // those batches and won't start executing the batches if the available budget
 // is insufficient. The function exits when an error is encountered by one of
@@ -1040,92 +1040,18 @@ func (w *workerCoordinator) performRequestAsync(
 				return
 			}
 
-			var resumeReq singleRangeBatch
-			// We will reuse the slices for the resume spans, if any.
-			resumeReq.reqs = req.reqs[:0]
-			resumeReq.positions = req.positions[:0]
-			var results []Result
-			var numCompleteGetResponses int
-			// memoryFootprintBytes tracks the total memory footprint of
-			// non-empty responses. This will be equal to the sum of memory
-			// tokens created for all Results.
-			var memoryFootprintBytes int64
-			var hasNonEmptyScanResponse bool
-			for i, resp := range br.Responses {
-				enqueueKey := req.positions[i]
-				if w.s.enqueueKeys != nil {
-					enqueueKey = w.s.enqueueKeys[req.positions[i]]
-				}
-				reply := resp.GetInner()
-				origReq := req.reqs[i]
-				// Unset the original request so that we lose the reference to
-				// the span.
-				req.reqs[i] = roachpb.RequestUnion{}
-				switch origRequest := origReq.GetInner().(type) {
-				case *roachpb.GetRequest:
-					get := reply.(*roachpb.GetResponse)
-					if get.ResumeSpan != nil {
-						// This Get wasn't completed - update the original
-						// request according to the ResumeSpan and include it
-						// into the batch again.
-						origRequest.SetSpan(*get.ResumeSpan)
-						resumeReq.reqs = append(resumeReq.reqs, origReq)
-						resumeReq.positions = append(resumeReq.positions, req.positions[i])
-					} else {
-						// This Get was completed.
-						toRelease := int64(get.Size())
-						result := Result{
-							GetResp: get,
-							// This currently only works because all requests
-							// are unique.
-							EnqueueKeysSatisfied: []int{enqueueKey},
-							position:             req.positions[i],
-						}
-						result.memoryTok.streamer = w.s
-						result.memoryTok.toRelease = toRelease
-						memoryFootprintBytes += toRelease
-						results = append(results, result)
-						numCompleteGetResponses++
-					}
-
-				case *roachpb.ScanRequest:
-					scan := reply.(*roachpb.ScanResponse)
-					resumeSpan := scan.ResumeSpan
-					if len(scan.Rows) > 0 || len(scan.BatchResponses) > 0 {
-						toRelease := int64(scan.Size())
-						result := Result{
-							// This currently only works because all requests
-							// are unique.
-							EnqueueKeysSatisfied: []int{enqueueKey},
-							position:             req.positions[i],
-						}
-						result.memoryTok.streamer = w.s
-						result.memoryTok.toRelease = toRelease
-						result.ScanResp.ScanResponse = scan
-						// Complete field will be set below.
-						memoryFootprintBytes += toRelease
-						results = append(results, result)
-						hasNonEmptyScanResponse = true
-					}
-					if resumeSpan != nil {
-						// This Scan wasn't completed - update the original
-						// request according to the resumeSpan and include it
-						// into the batch again.
-						origRequest.SetSpan(*resumeSpan)
-						resumeReq.reqs = append(resumeReq.reqs, origReq)
-						resumeReq.positions = append(resumeReq.positions, req.positions[i])
-					}
-				}
-			}
+			// First, we have to reconcile the memory budget. We do it
+			// separately from processing the results because we want to know
+			// how many Gets and Scans need to be allocated for the ResumeSpans,
+			// if any are present. At the moment, due to limitations of the KV
+			// layer (#75452) we cannot reuse original requests because the KV
+			// doesn't allow mutability.
+			memoryFootprintBytes, resumeReqsMemUsage, numIncompleteGets, numIncompleteScans := calculateFootprint(req, br)
 
 			// Now adjust the budget based on the actual memory footprint of
 			// non-empty responses as well as resume spans, if any.
 			respOverestimate := targetBytes - memoryFootprintBytes
-			var reqsMemUsage int64
-			if len(resumeReq.reqs) > 0 {
-				reqsMemUsage = requestsMemUsage(resumeReq.reqs)
-			}
-			reqOveraccounted := req.reqsReservedBytes - reqsMemUsage
+			reqOveraccounted := req.reqsReservedBytes - resumeReqsMemUsage
 			overaccountedTotal := respOverestimate + reqOveraccounted
 			if overaccountedTotal >= 0 {
 				w.s.budget.release(ctx, overaccountedTotal)
@@ -1169,8 +1095,6 @@ func (w *workerCoordinator) performRequestAsync(
 					return
 				}
 			}
-			// Update the resume request accordingly.
-			resumeReq.reqsReservedBytes = reqsMemUsage
 
 			// Do admission control after we've finalized the memory accounting.
 			if br != nil && w.responseAdmissionQ != nil {
@@ -1185,24 +1109,209 @@ func (w *workerCoordinator) performRequestAsync(
 				}
 			}
 
-			// If we have any results, finalize them.
-			if len(results) > 0 {
-				w.finalizeSingleRangeResults(
-					results, memoryFootprintBytes, hasNonEmptyScanResponse,
-					numCompleteGetResponses,
-				)
-			}
-
-			// If we have any incomplete requests, add them back into the work
-			// pool.
-			if len(resumeReq.reqs) > 0 {
-				w.addRequest(resumeReq)
-			}
+			// Finally, process the results and add the ResumeSpans to be
+			// processed as well.
+			w.processSingleRangeResults(
+				req, br, memoryFootprintBytes, resumeReqsMemUsage,
+				numIncompleteGets, numIncompleteScans,
+			)
 		}); err != nil {
 		// The new goroutine for the request wasn't spun up, so we have to
 		// perform the cleanup of this request ourselves.
 		w.asyncRequestCleanup()
 		w.s.setError(err)
+	}
+}
+
+// calculateFootprint calculates the memory footprint of the batch response as
+// well as of the requests that will have to be created for the ResumeSpans.
+// - memoryFootprintBytes tracks the total memory footprint of non-empty
+// responses. This will be equal to the sum of memory tokens created for all
+// Results.
+// - resumeReqsMemUsage tracks the memory usage of the requests for the
+// ResumeSpans.
+func calculateFootprint(
+	req singleRangeBatch, br *roachpb.BatchResponse,
+) (
+	memoryFootprintBytes int64,
+	resumeReqsMemUsage int64,
+	numIncompleteGets, numIncompleteScans int,
+) {
+	// Note that we cannot use Size() methods that are automatically generated
+	// by the protobuf library because they account for things differently from
+	// how the memory usage is accounted for by the KV layer for the purposes of
+	// tracking TargetBytes limit.
+
+	// getRequestScratch and scanRequestScratch are used to calculate
+	// the size of requests when we set the ResumeSpans on them.
+	var getRequestScratch roachpb.GetRequest
+	var scanRequestScratch roachpb.ScanRequest
+	for i, resp := range br.Responses {
+		reply := resp.GetInner()
+		switch req.reqs[i].GetInner().(type) {
+		case *roachpb.GetRequest:
+			get := reply.(*roachpb.GetResponse)
+			if get.ResumeSpan != nil {
+				// This Get wasn't completed.
+				getRequestScratch.SetSpan(*get.ResumeSpan)
+				resumeReqsMemUsage += int64(getRequestScratch.Size())
+				numIncompleteGets++
+			} else {
+				// This Get was completed.
+				memoryFootprintBytes += getResponseSize(get)
+			}
+		case *roachpb.ScanRequest:
+			scan := reply.(*roachpb.ScanResponse)
+			if len(scan.Rows) > 0 || len(scan.BatchResponses) > 0 {
+				memoryFootprintBytes += scanResponseSize(scan)
+			}
+			if scan.ResumeSpan != nil {
+				// This Scan wasn't completed.
+				scanRequestScratch.SetSpan(*scan.ResumeSpan)
+				resumeReqsMemUsage += int64(scanRequestScratch.Size())
+				numIncompleteScans++
+			}
+		}
+	}
+	// This addendum is the first step of requestsMemUsage() and we've already
+	// added the size of each resume request above.
+	resumeReqsMemUsage += requestUnionOverhead * int64(numIncompleteGets+numIncompleteScans)
+	return memoryFootprintBytes, resumeReqsMemUsage, numIncompleteGets, numIncompleteScans
+}
+
+// processSingleRangeResults creates a Result for each non-empty response found
+// in the BatchResponse. The ResumeSpans, if found, are added into a new
+// singleRangeBatch request that is added to be picked up by the mainLoop of the
+// worker coordinator. This method assumes that req is no longer needed by the
+// caller, so req.positions is reused for the ResumeSpans.
+//
+// It also assumes that the budget has already been reconciled with the
+// reservations for Results that will be created.
+func (w *workerCoordinator) processSingleRangeResults(
+	req singleRangeBatch,
+	br *roachpb.BatchResponse,
+	memoryFootprintBytes int64,
+	resumeReqsMemUsage int64,
+	numIncompleteGets, numIncompleteScans int,
+) {
+	numIncompleteRequests := numIncompleteGets + numIncompleteScans
+	var resumeReq singleRangeBatch
+	// We have to allocate the new slice for requests, but we can reuse the
+	// positions slice.
+	resumeReq.reqs = make([]roachpb.RequestUnion, numIncompleteRequests)
+	// numIncompleteRequests will never exceed the number of requests in req.
+	resumeReq.positions = req.positions[:numIncompleteRequests]
+	// We've already reconciled the budget with the actual reservation for the
+	// requests with the ResumeSpans.
+	resumeReq.reqsReservedBytes = resumeReqsMemUsage
+	gets := make([]struct {
+		req   roachpb.GetRequest
+		union roachpb.RequestUnion_Get
+	}, numIncompleteGets)
+	scans := make([]struct {
+		req   roachpb.ScanRequest
+		union roachpb.RequestUnion_Scan
+	}, numIncompleteScans)
+	var results []Result
+	var numCompleteGetResponses int
+	var hasNonEmptyScanResponse bool
+	var resumeReqIdx int
+	// memoryTokensBytes accumulates all reservations that are made for all
+	// Results created below. The accounting for these reservations has already
+	// been performed, and memoryTokensBytes should be exactly equal to
+	// memoryFootprintBytes, so we use it only as an additional check.
+	var memoryTokensBytes int64
+	for i, resp := range br.Responses {
+		enqueueKey := req.positions[i]
+		if w.s.enqueueKeys != nil {
+			enqueueKey = w.s.enqueueKeys[req.positions[i]]
+		}
+		reply := resp.GetInner()
+		switch origRequest := req.reqs[i].GetInner().(type) {
+		case *roachpb.GetRequest:
+			get := reply.(*roachpb.GetResponse)
+			if get.ResumeSpan != nil {
+				// This Get wasn't completed - update the original
+				// request according to the ResumeSpan and include it
+				// into the batch again.
+				newGet := gets[0]
+				gets = gets[1:]
+				newGet.req.SetSpan(*get.ResumeSpan)
+				newGet.req.KeyLocking = origRequest.KeyLocking
+				newGet.union.Get = &newGet.req
+				resumeReq.reqs[resumeReqIdx].Value = &newGet.union
+				resumeReq.positions[resumeReqIdx] = req.positions[i]
+				resumeReqIdx++
+			} else {
+				// This Get was completed.
+				result := Result{
+					GetResp: get,
+					// This currently only works because all requests
+					// are unique.
+					EnqueueKeysSatisfied: []int{enqueueKey},
+					position:             req.positions[i],
+				}
+				result.memoryTok.streamer = w.s
+				result.memoryTok.toRelease = getResponseSize(get)
+				memoryTokensBytes += result.memoryTok.toRelease
+				results = append(results, result)
+				numCompleteGetResponses++
+			}
+
+		case *roachpb.ScanRequest:
+			scan := reply.(*roachpb.ScanResponse)
+			if len(scan.Rows) > 0 || len(scan.BatchResponses) > 0 {
+				result := Result{
+					// This currently only works because all requests
+					// are unique.
+					EnqueueKeysSatisfied: []int{enqueueKey},
+					position:             req.positions[i],
+				}
+				result.memoryTok.streamer = w.s
+				result.memoryTok.toRelease = scanResponseSize(scan)
+				memoryTokensBytes += result.memoryTok.toRelease
+				result.ScanResp.ScanResponse = scan
+				// Complete field will be set below.
+				results = append(results, result)
+				hasNonEmptyScanResponse = true
+			}
+			if scan.ResumeSpan != nil {
+				// This Scan wasn't completed - update the original
+				// request according to the ResumeSpan and include it
+				// into the batch again.
+				newScan := scans[0]
+				scans = scans[1:]
+				newScan.req.SetSpan(*scan.ResumeSpan)
+				newScan.req.KeyLocking = origRequest.KeyLocking
+				newScan.union.Scan = &newScan.req
+				resumeReq.reqs[resumeReqIdx].Value = &newScan.union
+				resumeReq.positions[resumeReqIdx] = req.positions[i]
+				resumeReqIdx++
+			}
+		}
+	}
+
+	if buildutil.CrdbTestBuild {
+		if memoryFootprintBytes != memoryTokensBytes {
+			panic(errors.AssertionFailedf(
+				"different calculation of memory footprint\ncalculateFootprint: %d bytes\n"+
+					"processSingleRangeResults: %d bytes", memoryFootprintBytes, memoryTokensBytes,
+			))
+		}
+	}
+
+	// If we have any results, finalize them.
+	if len(results) > 0 {
+		w.finalizeSingleRangeResults(
+			results, memoryFootprintBytes, hasNonEmptyScanResponse,
+			numCompleteGetResponses,
+		)
+	}
+
+	// If we have any incomplete requests, add them back into the work
+	// pool.
+	if len(resumeReq.reqs) > 0 {
+		w.addRequest(resumeReq)
 	}
 }
 
@@ -1270,13 +1379,31 @@ func init() {
 	zeroIntSlice = make([]int, 1<<10)
 }
 
-const requestUnionSliceOverhead = int64(unsafe.Sizeof([]roachpb.RequestUnion{}))
+const requestUnionOverhead = int64(unsafe.Sizeof(roachpb.RequestUnion{}))
 
 func requestsMemUsage(reqs []roachpb.RequestUnion) int64 {
-	memUsage := requestUnionSliceOverhead
-	// Slice up to the capacity to account for everything.
-	for _, r := range reqs[:cap(reqs)] {
+	// RequestUnion.Size() ignores the overhead of RequestUnion object, so we'll
+	// account for it separately first.
+	memUsage := requestUnionOverhead * int64(cap(reqs))
+	// No need to account for elements past len(reqs) because those must be
+	// unset and we have already accounted for RequestUnion object above.
+	for _, r := range reqs {
 		memUsage += int64(r.Size())
 	}
 	return memUsage
+}
+
+// getResponseSize calculates the size of the GetResponse similar to how it is
+// accounted for TargetBytes parameter by the KV layer.
+func getResponseSize(get *roachpb.GetResponse) int64 {
+	if get.Value == nil {
+		return 0
+	}
+	return int64(len(get.Value.RawBytes))
+}
+
+// scanResponseSize calculates the size of the ScanResponse similar to how it is
+// accounted for TargetBytes parameter by the KV layer.
+func scanResponseSize(scan *roachpb.ScanResponse) int64 {
+	return scan.NumBytes
 }

--- a/pkg/kv/kvclient/kvstreamer/streamer_test.go
+++ b/pkg/kv/kvclient/kvstreamer/streamer_test.go
@@ -12,6 +12,7 @@ package kvstreamer
 
 import (
 	"context"
+	"fmt"
 	"math"
 	"testing"
 
@@ -28,6 +29,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/util/log"
 	"github.com/cockroachdb/cockroach/pkg/util/mon"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
+	"github.com/dustin/go-humanize"
 	"github.com/stretchr/testify/require"
 )
 
@@ -268,4 +270,69 @@ func TestStreamerBudgetErrorInEnqueue(t *testing.T) {
 		reqs[1] = makeGetRequest(limitBytes/2 + 1)
 		require.Error(t, streamer.Enqueue(ctx, reqs, nil /* enqueueKeys */))
 	})
+}
+
+// TestStreamerCorrectlyDiscardsResponses verifies that the Streamer behaves
+// correctly in a scenario when partial results are returned, but the original
+// memory reservation was under-provisioned and the budget cannot be reconciled,
+// so the responses have to be discarded.
+func TestStreamerCorrectlyDiscardsResponses(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// Start a cluster with large --max-sql-memory parameter so that the
+	// Streamer isn't hitting the root budget exceeded error.
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{
+		SQLMemoryPoolSize: 1 << 30, /* 1GiB */
+	})
+	ctx := context.Background()
+	defer s.Stopper().Stop(ctx)
+
+	// The initial estimate for TargetBytes argument for each asynchronous
+	// request by the Streamer will be numRowsPerRange x initialAvgResponseSize,
+	// so we pick the blob size such that about half of rows are included in the
+	// partial responses.
+	const blobSize = 2 * initialAvgResponseSize
+	const numRows = 20
+	const numRowsPerRange = 4
+
+	_, err := db.Exec("CREATE TABLE t (pk INT PRIMARY KEY, k INT, blob STRING, INDEX (k))")
+	require.NoError(t, err)
+	for i := 0; i < numRows; i++ {
+		if i > 0 && i%numRowsPerRange == 0 {
+			// Create a new range for the next numRowsPerRange rows.
+			_, err = db.Exec(fmt.Sprintf("ALTER TABLE t SPLIT AT VALUES(%d)", i))
+			require.NoError(t, err)
+		}
+		_, err = db.Exec(fmt.Sprintf("INSERT INTO t SELECT %d, 1, repeat('a', %d)", i, blobSize))
+		require.NoError(t, err)
+	}
+
+	// Populate the range cache.
+	_, err = db.Exec("SELECT count(*) from t")
+	require.NoError(t, err)
+
+	// Perform an index join to read the blobs.
+	query := "SELECT sum(length(blob)) FROM t@t_k_idx WHERE k = 1"
+	// Use several different workmem limits to exercise somewhat different
+	// scenarios.
+	//
+	// All of these values allow for all initial requests to be issued
+	// asynchronously, but only for some of the responses to be "accepted" by
+	// the budget. This includes 4/3 factor since the vectorized ColIndexJoin
+	// gives 3/4 of the workmem limit to the Streamer.
+	for _, workmem := range []int{
+		3 * initialAvgResponseSize * numRows / 2,
+		7 * initialAvgResponseSize * numRows / 4,
+		2 * initialAvgResponseSize * numRows,
+	} {
+		t.Run(fmt.Sprintf("workmem=%s", humanize.Bytes(uint64(workmem))), func(t *testing.T) {
+			_, err = db.Exec(fmt.Sprintf("SET distsql_workmem = '%dB'", workmem))
+			require.NoError(t, err)
+			row := db.QueryRow(query)
+			var sum int
+			require.NoError(t, row.Scan(&sum))
+			require.Equal(t, numRows*blobSize, sum)
+		})
+	}
 }

--- a/pkg/sql/mem_limit_test.go
+++ b/pkg/sql/mem_limit_test.go
@@ -12,6 +12,9 @@ package sql
 
 import (
 	"context"
+	"fmt"
+	"regexp"
+	"strconv"
 	"strings"
 	"sync"
 	"testing"
@@ -129,4 +132,68 @@ func TestMemoryLimit(t *testing.T) {
 			})
 		})
 	}
+}
+
+// TestStreamerTightBudget verifies that the Streamer utilizes its available
+// budget as tightly as possible, without incurring unnecessary debt. It gives
+// the Streamer such a budget that a single result puts it in debt, so there
+// should be no more than a single request "in progress" (i.e. one request in
+// flight or one unreleased result).
+func TestStreamerTightBudget(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	// Start a cluster with large --max-sql-memory parameter so that the
+	// Streamer isn't hitting the root budget exceeded error.
+	s, db, _ := serverutils.StartServer(t, base.TestServerArgs{
+		SQLMemoryPoolSize: 1 << 30, /* 1GiB */
+	})
+	ctx := context.Background()
+	defer s.Stopper().Stop(ctx)
+
+	const blobSize = 1 << 20
+	const numRows = 5
+
+	_, err := db.Exec("CREATE TABLE t (pk INT PRIMARY KEY, k INT, blob STRING, INDEX (k))")
+	require.NoError(t, err)
+	for i := 0; i < numRows; i++ {
+		if i > 0 {
+			// Create a new range for this row.
+			_, err = db.Exec(fmt.Sprintf("ALTER TABLE t SPLIT AT VALUES(%d)", i))
+			require.NoError(t, err)
+		}
+		_, err = db.Exec(fmt.Sprintf("INSERT INTO t SELECT %d, 1, repeat('a', %d)", i, blobSize))
+		require.NoError(t, err)
+	}
+
+	// Populate the range cache.
+	_, err = db.Exec("SELECT count(*) from t")
+	require.NoError(t, err)
+
+	// Set the workmem limit to a low value such that it will allow the Streamer
+	// to have at most one request to be "in progress".
+	_, err = db.Exec(fmt.Sprintf("SET distsql_workmem = '%dB'", blobSize))
+	require.NoError(t, err)
+
+	// Perform an index join to read the blobs.
+	query := "EXPLAIN ANALYZE SELECT sum(length(blob)) FROM t@t_k_idx WHERE k = 1"
+	maximumMemoryUsageRegex := regexp.MustCompile(`maximum memory usage: (\d+\.\d+) MiB`)
+	rows, err := db.QueryContext(ctx, query)
+	require.NoError(t, err)
+	for rows.Next() {
+		var res string
+		require.NoError(t, rows.Scan(&res))
+		if matches := maximumMemoryUsageRegex.FindStringSubmatch(res); len(matches) > 0 {
+			usage, err := strconv.ParseFloat(matches[1], 64)
+			require.NoError(t, err)
+			// We expect that the maximum memory usage is about 2MiB (1MiB is
+			// accounted for by the Streamer, and another 1MiB is accounted for
+			// by the ColIndexJoin when the blob is copied into the columnar
+			// batch). We allow for 0.1MiB for other memory usage in the query.
+			maxAllowed := 2.1
+			require.GreaterOrEqual(t, maxAllowed, usage, "unexpectedly high memory usage")
+			return
+		}
+	}
+	t.Fatal("unexpectedly didn't find a match for maximum memory usage")
 }

